snowflakedb/snowflake-cli
# Workspaces application package bundle

### Pre-review checklist
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * I've described my changes in the release notes.

### Changes description

Implementing the v2 `bundle()` logic of `ApplicationPackageEntity`, enabling `snow ws bundle --entity-id=my_app_pkg`. This is a reuse of the existing v1 `snow app bundle`.

Due to circular dependencies, this change required extracting some of the existing code to dedicated modules:
* `OnErrorType`
* `Variable`
* `parse_key_value_variables`
* `print_diff_to_console`

Additionally, wrapped one of the imports in `cli_global_context.py` with `TYPE_CHECKING`, as breaking this cycle required a significant refactor. See PR comment on the file.

### Manual test

Running `snow ws bundle` with a simple v2 definition, verifying that the deploy directory is created with the correct artifacts:
```
$ snow ws bundle --entity-id=pkg

Bundle generated at /tmp/pdfv2-test/output/deploy
```
